### DISCUSSION TITLE
Are there negation types/operators?

### DISCUSSION THREAD

**Discussion by sschneider-ihre-pvs** [colinhacks/zod#926] [Q&A] [answered]
Is there something like `.xor()` or `.not()` possible?
For example `z.string().optional().xor(z.string().nonempty())` so either it does not have a value at all or if it has a value it ought to be not empty. For example if you verify a payload that only has a changeset and this value was not changed hence not in the changeset but if it was changed than the value is not allowed to be empty.

**Comment by JacobWeisenburger** [comment, marked as answer]
There isn't a concept of `xor` or `not` that I know of, but you could do your example as shown below:

```ts
const optionalNonEmptyString = z.string().min( 1 ).optional()
type OptionalNonEmptyString = z.infer<typeof optionalNonEmptyString>
// type OptionalNonEmptyString = string | undefined

console.log( optionalNonEmptyString.safeParse( 'foo' ).success ) // true
console.log( optionalNonEmptyString.safeParse( undefined ).success ) // true
console.log( optionalNonEmptyString.safeParse( '' ).success ) // false
```

For other schemas that you want to use not, you could do something using [refine](https://github.com/colinhacks/zod#refine).
